### PR TITLE
Fix MAGN-5768 With a "dirty definition" closing Revit fails to prompt Dynamo for File Save

### PR DESCRIPTION
**Pull request by Randy-Ma**
### Purpose

This PR fixes MAGN-5768. When Dynamo is closed because of the host application is closed, the closing handler of Dynamo window won't be called. The result of this is there will be no chance to prompt users to whether or not save the Dynamo documents.

The fix here is try to check whether PerformShutdownSequence of the view model has been called successfully when the closed handler is called. If not, it will be called first.

### Declarations

- [x] The code base is in a better state after this PR
- [x] The level of testing this PR includes is appropriate
(I have run manual test.)

### Reviewers

@pboyer 

### FYIs

@riteshchandawar

### CODE REVIEW

**Comment by pboyer** [comment]
@Randy-Ma This is a clever fix that has evaded us for a while!  However, I want to know, is there a way to build an automated test for this?  As we can build a DynamoView programmatically, it seems possible.  Do you think you could add that?

**Comment by Randy-Ma** [comment]
@pboyer 
Thanks for the comments! The reason that I have not added an automatic test is that this test case has to be run with a host application, which is Revit for this defect. This means we need an API to close Revit and then check whether the prompt to save dialog comes out. I have not seen a straightforward way to handle these.

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs [hunk 1]
```diff
@@ -46,6 +46,7 @@ public partial class DynamoView : Window
         private Stopwatch _timer;
         private StartPageViewModel startPage;
         private int tabSlidingWindowStart, tabSlidingWindowEnd;
+        private bool isPSSCalledOnViewModelNoCancel = false;
```
Comment (pboyer) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs]: Let's add a comment to clarify the purpose of this flag.

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs [hunk 2]
```diff
@@ -738,16 +739,32 @@ private void WindowClosing(object sender, CancelEventArgs e)
...
             }
         }
 
+        private void WindowClosing(object sender, CancelEventArgs e)
+        {
+            if (!PerformShutdownSequenceOnViewModel() && !DynamoModel.IsTestMode)
+                e.Cancel = true;
+            else
+                isPSSCalledOnViewModelNoCancel = true;
+        }
```
Comment (pboyer) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs]: This is just style, but I'd prefer if you used curly brackets around these code blocks.  This makes it harder to make silly refactoring mistakes.

**Comment by pboyer** [comment]
Ok, that's fine @Randy-Ma.  If there's really no way to test it, then that's okay, but I fear this may come back to haunt us.  :(  

I've suggested a few minor changes here.  Once those changes are in and you get the green light from Travis, feel free to merge in.

**Comment by Randy-Ma** [comment]
@pboyer 
Thanks for the suggestions! I have made some changes according to your comments.

**Comment by jnealb** [comment]
@pboyer @Randy-Ma This is part of the backlog of tests to be implemented with an external test tool.

**Pull request merged**